core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -1487,7 +1488,7 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
   def unsafe: UnsafeAPI =
     new UnsafeAPI {
       def addObserver(observer: Exit[E, A] => Unit)(implicit unsafe: Unsafe): Unit =
-        self.addObserver(observer)
+        self.tell(FiberMessage.Stateful(_.asInstanceOf[FiberRuntime[E, A]].addObserver(observer)))
```

Comment:
Perhaps let's avoid the message passing and restarting of the fiber when the _exitValue is already set:

```suggestion
        val exit = _exitValue
        if (exit ne null) observer(exit)
        else self.tell(FiberMessage.Stateful(_.asInstanceOf[FiberRuntime[E, A]].addObserver(observer)))
```

Reply:
Also, do we need the `asInstanceOf`? Doesn't seem to me like it's needed

Reply:
> There is a comment in there that implies that notification order is important.

I think notification order is only required when we need them to be strictly ordered (i.e., if I add 2 observers one after the other then I want to make sure they're called in the right order). But for other observers we really shouldn't care, so it's OK to do that.

Also, we've been doing this exact thing so far and we didn't have any issues with ordering, so I highly doubt it'll be an issue.

Reply:
> his does something similar:

I was going to mention this, can you also change that to use `unsafe.addObserver` now?